File: android/core/engine/src/test/java/com/xianxia/sect/core/engine/domain/disciple/DiscipleFactoryTest.kt
package com.xianxia.sect.core.engine.domain.disciple

import com.xianxia.sect.core.model.SocialData
import com.xianxia.sect.core.util.NameService
import org.junit.Assert.*
import org.junit.Test

/**
 * 验证 DiscipleFactory 统一构造 — 三处构造站点（recruitDisciple /
 * refreshRecruitList / createChild）通过不同 DiscipleSeed 得到一致的
 * 六段逻辑（variance / comprehension / skills / baseStats / lifespan /
 * talentIds）。
 */
class DiscipleFactoryTest {

    private val factory = DiscipleFactory()

    // ---- 辅助函数 ----

    private fun newSeed(
        id: String = "test-001",
        gender: String = "男",
        spiritRootType: String = "火",
        age: Int = 18,
        realmLayer: Int = 1,
        random: kotlin.random.Random = kotlin.random.Random(42)
    ): DiscipleFactory.DiscipleSeed {
        return DiscipleFactory.DiscipleSeed(
            id = id,
            gender = gender,
            nameResult = NameService.NameResult("测试", "弟子"),
            spiritRootType = spiritRootType,
            age = age,
            realmLayer = realmLayer,
            social = SocialData(),
            nextInt = { from, _ -> from }, // 确定性：总是取最小值
            random = random
        )
    }

    // ---- 方差 / 悟性 / 技能 ----

    @Test
    fun `create - variance is deterministic given nextInt`() {
        val d = factory.create(newSeed())
        // nextInt 固定返回 from → Box-Muller z≈4.291 → 截断至 50
        assertEquals(50, d.combat.hpVariance)
        assertEquals(50, d.combat.mpVariance)
        assertEquals(50, d.combat.physicalAttackVariance)
        assertEquals(50, d.combat.magicAttackVariance)
        assertEquals(50, d.combat.physicalDefenseVariance)
        assertEquals(50, d.combat.magicDefenseVariance)
        assertEquals(50, d.combat.speedVariance)
    }

    @Test
    fun `create - skills are deterministic given nextInt`() {
        val d = factory.create(newSeed())
        // 同 gaussianInt 逻辑：u1=0.0001, u2=0.0, z≈4.291
        // skill = round(4.291*16.5 + 50.5) = round(121.3) = 121
        // 2026-08-12 上限 100→200：121 不再截断（原断言 100）；loyalty 上限 100 不变仍截断
        assertEquals(121, d.skills.intelligence)
        assertEquals(121, d.skills.charm)
        assertEquals(100, d.skills.loyalty)
        assertEquals(121, d.skills.morality)
        assertEquals(121, d.skills.artifactRefining)
        assertEquals(121, d.skills.pillRefining)
        assertEquals(121, d.skills.spiritPlanting)
        assertEquals(121, d.skills.mining)
        assertEquals(121, d.skills.teaching)
    }

    @Test
    fun `create - single spirit root yields high comprehension`() {
        val d = factory.create(newSeed(spiritRootType = "火"))
        // nextInt 固定 from=80，单灵根 from=80 → comprehension=80
        assertEquals(80, d.skills.comprehension)
    }

    @Test
    fun `create - two spirit roots yield mid comprehension`() {
        val d = factory.create(newSeed(spiritRootType = "火,水"))
        // nextInt 固定 from=60，双灵根 from=60 → comprehension=60
        assertEquals(60, d.skills.comprehension)
    }

    @Test
    fun `create - three spirit roots yield lower comprehension`() {
        val d = factory.create(newSeed(spiritRootType = "火,水,木"))
        assertEquals(40, d.skills.comprehension)
    }

    @Test
    fun `create - four spirit roots yield minimal comprehension`() {
        val d = factory.create(newSeed(spiritRootType = "火,水,木,金"))
        assertEquals(20, d.skills.comprehension)
    }

    @Test
    fun `create - five spirit roots yield worst comprehension`() {
        val d = factory.create(newSeed(spiritRootType = "火,水,木,金,土"))
        assertEquals(1, d.skills.comprehension)
    }

    // ---- 资质（2026-08-12 新增固定属性，与悟性同阶梯）----

    @Test
    fun `create - aptitude ladder mirrors comprehension per root count`() {
        // nextInt 固定返回 from：1根→80、2根→60、3根→40、4根→20、5根→1（同悟性阶梯）
        assertEquals(80, factory.create(newSeed(spiritRootType = "火")).skills.aptitude)
        assertEquals(60, factory.create(newSeed(spiritRootType = "火,水")).skills.aptitude)
        assertEquals(40, factory.create(newSeed(spiritRootType = "火,水,木")).skills.aptitude)
        assertEquals(20, factory.create(newSeed(spiritRootType = "火,水,木,金")).skills.aptitude)
        assertEquals(1, factory.create(newSeed(spiritRootType = "火,水,木,金,土")).skills.aptitude)
    }

    @Test
    fun `create - aptitude generation avoids sentinel 50`() {
        // nextInt 在资质 5 根段（from=1, until=201）命中哨兵 50 → 生成强制 +1 收敛为 51，
        // 否则资质==50 会被读档自愈误判为"未生成"重复重算（资质跳变）。
        // 其他段（数组索引如 PortraitPool）返回 from 保持安全。
        val seed = newSeed(spiritRootType = "火,水,木,金,土").copy(
            nextInt = { from, until -> if (from == 1 && until == 201) 50 else from }
        )
        assertEquals(51, factory.create(seed).skills.aptitude)
    }

    @Test
    fun `create - aptitude range respects SKILL_MAX 200`() {
        // 全值域抽样：资质（与悟性同区间 [min, 200]）恒 ≤ 200
        repeat(50) { i ->
            val d = factory.create(newSeed(id = "range-$i"))
            assertTrue("资质应 ≤ 200，实际 ${d.skills.aptitude}", d.skills.aptitude <= 200)
            assertTrue("资质应 ≥ 1，实际 ${d.skills.aptitude}", d.skills.aptitude >= 1)
        }
    }

    // ---- 基础属性（calculateBaseStatsWithVariance） ----

    @Test
    fun `create - baseStats are populated`() {
        val d = factory.create(newSeed())
        assertTrue("baseHp should be > 0", d.combat.baseHp > 0)
        assertTrue("baseMp should be > 0", d.combat.baseMp > 0)
        assertTrue("basePhysicalAttack should be > 0", d.combat.basePhysicalAttack > 0)
        assertTrue("baseMagicAttack should be > 0", d.combat.baseMagicAttack > 0)
        assertTrue("basePhysicalDefense should be > 0", d.combat.basePhysicalDefense > 0)
        assertTrue("baseMagicDefense should be > 0", d.combat.baseMagicDefense > 0)
        assertTrue("baseSpeed should be > 0", d.combat.baseSpeed > 0)
    }

    // ---- 寿命 ----

    @Test
    fun `create - lifespan is positive`() {
        val d = factory.create(newSeed())
        assertTrue("lifespan should be > 0", d.lifespan > 0)
    }

    // ---- 天赋 ----

    @Test
    fun `create - disciple has valid id`() {
        val d = factory.create(newSeed())
        assertTrue("id should not be blank", d.id.isNotBlank())
        assertEquals("男", d.gender)
        assertEquals(18, d.age)
    }

    // ---- 三站一致性 ----

    @Test
    fun `create - deterministic fields match given seed`() {
        val seed = newSeed()
        val d1 = factory.create(seed)
        val d2 = factory.create(seed)
        // nextInt 确定性字段
        assertEquals(d1.gender, d2.gender)
        assertEquals(d1.age, d2.age)
        assertEquals(d1.realm, d2.realm)
        assertEquals(d1.realmLayer, d2.realmLayer)
        assertEquals(d1.spiritRootType, d2.spiritRootType)
        assertEquals(d1.skills.comprehension, d2.skills.comprehension)
        assertEquals(d1.combat.hpVariance, d2.combat.hpVariance)
    }

    @Test
    fun `create - same seed produces identical disciple including traits`() {
        // 两个独立构造的相同 seed（各持 Random(42)）→ 序列一致 → 结果完全一致
        val d1 = factory.create(newSeed())
        val d2 = factory.create(newSeed())
        assertEquals(d1.talentIds, d2.talentIds)
        assertEquals(d1.physiqueIds, d2.physiqueIds)
        assertEquals(d1.affixIds, d2.affixIds)
        assertEquals(d1.name, d2.name)
    }

    @Test
    fun `create - different seeds produce different disciples`() {
        val d1 = factory.create(newSeed(id = "a", spiritRootType = "火"))
        val d2 = factory.create(newSeed(id = "b", spiritRootType = "火,水"))
        assertNotEquals(d1.skills.comprehension, d2.skills.comprehension)
    }

    // ---- 边界 ----

    @Test
    fun `create - female gender`() {
        val d = factory.create(newSeed(gender = "女"))
        assertEquals("女", d.gender)
    }

    @Test
    fun `create - age is preserved`() {
        val d = factory.create(newSeed(age = 25))
        assertEquals(25, d.age)
    }

    @Test
    fun `create - realmLayer is preserved`() {
        val d = factory.create(newSeed(realmLayer = 3))
        assertEquals(3, d.realmLayer)
    }

    // ---- 分布验证 ----

    @Test
    fun `create - variance center near zero with real random`() {
        val values = mutableListOf<Int>()
        val kotlinRng = kotlin.random.Random
        repeat(1000) {
            val seed = DiscipleFactory.DiscipleSeed(
                id = "dist-test-$it",
                gender = "男",
                nameResult = NameService.NameResult("测试", "弟子"),
                spiritRootType = "火",
                age = 18,
                realmLayer = 1,
                social = SocialData(),
                nextInt = { from, until -> from + kotlinRng.nextInt(until - from) },
                random = kotlinRng
            )
            values.add(factory.create(seed).combat.hpVariance)
        }
        val mean = values.average()
        // 均值应接近0（方差[-50,50]的正态分布）
        assertTrue("Variance mean should be near 0: $mean", mean in -10.0..10.0)
        // 至少70%的值落在[-30,30]范围内（约2-sigma）
        val withinMid = values.count { it in -30..30 }
        assertTrue("Less than 70% in [-30,30]: ${withinMid}/1000", withinMid >= 700)
    }
}

File: android/app/src/main/cpp/gamecore/test/disciple_factory_test.cpp
// ============================================================
// disciple_factory_test.cpp — 弟子创建黄金序列
//
// 确定性守护：固定种子 + 固定消费序 ⇒ 固定输出。黄金值来源：
// Kotlin `DiscipleFactory.create`（真相源）经 DiffDiscipleFactoryTest
// （JNI 对拍，同种子逐字段位级一致）确认后固化——本文件防 C++ 侧
// 回归漂移，正确性锚定在 Kotlin 对拍测试。
//
// 另含分布统计断言：数量 0-5 / 品阶四档 / 资质哨兵 50 规避 / 技能上限。
// ============================================================
#include <gtest/gtest.h>

#include <algorithm>
#include <cstdint>
#include <iostream>
#include <set>
#include <string>
#include <vector>

#include "gamecore/rng/pcg_xsh_rr.h"
#include "gamecore/system/disciple_factory.h"

namespace {

using gamecore::rng::DeterministicRng;
using gamecore::system::DiscipleCreationSeed;
using gamecore::system::createDisciple;

DiscipleCreationSeed kSeed() {
    DiscipleCreationSeed s;
    s.id = "golden-001";
    s.gender = "male";
    s.fullName = "李逍遥";
    s.surname = "李";
    s.spiritRootType = "火";
    s.age = 18;
    s.realm = 9;
    s.realmLayer = 1;
    return s;
}

TEST(DiscipleFactory, GoldenSequenceSeed42) {
    auto rng = DeterministicRng::fromSeed(42);
    const auto d = createDisciple(kSeed(), rng);
    // 黄金值（Kotlin DiscipleFactory.create 同种子输出，DiffDiscipleFactoryTest
    // 对拍逐字段确认后固化——seed=42/male/单灵根火）
    EXPECT_EQ("male_disciple_14", d.portraitRes);
    EXPECT_EQ(21, d.hpVariance);
    EXPECT_EQ(18, d.mpVariance);
    EXPECT_EQ(10, d.physicalAttackVariance);
    EXPECT_EQ(16, d.magicAttackVariance);
    EXPECT_EQ(8, d.physicalDefenseVariance);
    EXPECT_EQ(6, d.magicDefenseVariance);
    EXPECT_EQ(8, d.speedVariance);
    EXPECT_EQ(82, d.comprehension);
    EXPECT_EQ(95, d.aptitude);
    EXPECT_EQ(45, d.intelligence);
    EXPECT_EQ(32, d.charm);
    EXPECT_EQ(43, d.loyalty);
    EXPECT_EQ(48, d.morality);
    EXPECT_EQ(43, d.artifactRefining);
    EXPECT_EQ(33, d.pillRefining);
    EXPECT_EQ(66, d.spiritPlanting);
    EXPECT_EQ(51, d.mining);
    EXPECT_EQ(81, d.teaching);
    EXPECT_EQ(145, d.baseHp);
    EXPECT_EQ(70, d.baseMp);
    EXPECT_EQ(13, d.basePhysicalAttack);
    EXPECT_EQ(13, d.baseMagicAttack);
    EXPECT_EQ(10, d.basePhysicalDefense);
    EXPECT_EQ(8, d.baseMagicDefense);
    EXPECT_EQ(16, d.baseSpeed);
    EXPECT_EQ(80, d.lifespan);
    EXPECT_TRUE(d.talentIds.empty());
    EXPECT_EQ(std::vector<std::string>({"neg_phys_defense", "r1_phys_cult_speed"}),
              d.physiqueIds);
    EXPECT_TRUE(d.affixIds.empty());
}

TEST(DiscipleFactory, GoldenSequenceSeed987654321Female) {
    auto rng = DeterministicRng::fromSeed(987654321);
    DiscipleCreationSeed s = kSeed();
    s.id = "golden-002";
    s.gender = "female";
    s.fullName = "慕容雪";
    s.surname = "慕容";
    s.spiritRootType = "火,水,木,金,土";
    const auto d = createDisciple(s, rng);
    // 黄金值（Kotlin DiscipleFactory.create 同种子输出，DiffDiscipleFactoryTest
    // 对拍逐字段确认后固化——seed=987654321/female/五灵根，覆盖负面体质抽取
    // 与三分类多特质路径）
    EXPECT_EQ("female_disciple_8", d.portraitRes);
    EXPECT_EQ(-18, d.hpVariance);
    EXPECT_EQ(3, d.mpVariance);
    EXPECT_EQ(-9, d.physicalAttackVariance);
    EXPECT_EQ(-9, d.magicAttackVariance);
    EXPECT_EQ(-28, d.physicalDefenseVariance);
    EXPECT_EQ(-9, d.magicDefenseVariance);
    EXPECT_EQ(4, d.speedVariance);
    EXPECT_EQ(17, d.comprehension);
    EXPECT_EQ(5, d.aptitude);
    EXPECT_EQ(36, d.intelligence);
    EXPECT_EQ(58, d.charm);
    EXPECT_EQ(77, d.loyalty);
    EXPECT_EQ(59, d.morality);
    EXPECT_EQ(31, d.artifactRefining);
    EXPECT_EQ(54, d.pillRefining);
    EXPECT_EQ(5, d.spiritPlanting);
    EXPECT_EQ(45, d.mining);
    EXPECT_EQ(48, d.teaching);
    EXPECT_EQ(98, d.baseHp);
    EXPECT_EQ(61, d.baseMp);
    EXPECT_EQ(10, d.basePhysicalAttack);
    EXPECT_EQ(10, d.baseMagicAttack);
    EXPECT_EQ(7, d.basePhysicalDefense);
    EXPECT_EQ(7, d.baseMagicDefense);
    EXPECT_EQ(15, d.baseSpeed);
    EXPECT_EQ(80, d.lifespan);
    EXPECT_TRUE(d.talentIds.empty());
    EXPECT_EQ(std::vector<std::string>({"r2_phys_defense", "neg_phys_cult"}),
              d.physiqueIds);
    EXPECT_EQ(std::vector<std::string>({"r1_aff_pos_vice_sect_master",
                                        "r1_aff_pos_inner_elder",
                                        "r1_aff_cult_speed"}),
              d.affixIds);
}

TEST(DiscipleFactory, DeterministicAcrossInstances) {
    // 同种子重放 ⇒ 逐字段一致（防静态初始化序/未定义行为导致漂移）
    auto rngA = DeterministicRng::fromSeed(2024);
    auto rngB = DeterministicRng::fromSeed(2024);
    const auto a = createDisciple(kSeed(), rngA);
    const auto b = createDisciple(kSeed(), rngB);
    EXPECT_EQ(a.portraitRes, b.portraitRes);
    EXPECT_EQ(a.hpVariance, b.hpVariance);
    EXPECT_EQ(a.speedVariance, b.speedVariance);
    EXPECT_EQ(a.comprehension, b.comprehension);
    EXPECT_EQ(a.aptitude, b.aptitude);
    EXPECT_EQ(a.intelligence, b.intelligence);
    EXPECT_EQ(a.teaching, b.teaching);
    EXPECT_EQ(a.baseHp, b.baseHp);
    EXPECT_EQ(a.baseSpeed, b.baseSpeed);
    EXPECT_EQ(a.lifespan, b.lifespan);
    EXPECT_EQ(a.talentIds, b.talentIds);
    EXPECT_EQ(a.physiqueIds, b.physiqueIds);
    EXPECT_EQ(a.affixIds, b.affixIds);
}

TEST(DiscipleFactory, DistributionInvariants) {
    // 统计不变式：数量 0-5、技能 1-200（忠诚 ≤100）、方差 -50..50、
    // 资质避开哨兵 50、悟性/资质 1-200、lifespan ≥ 1
    std::set<int32_t> counts;
    std::set<int32_t> aptitudes;
    std::set<int32_t> comprehensions;
    for (int32_t i = 0; i < 500; ++i) {
        auto rng = DeterministicRng::fromSeed(9000 + i);
        DiscipleCreationSeed s = kSeed();
        s.id = "invariant-" + std::to_string(i);
        s.spiritRootType = (i % 5 == 0) ? "火" : "火,水,木,金,土";
        const auto d = createDisciple(s, rng);
        counts.insert(static_cast<int32_t>(d.talentIds.size()));
        aptitudes.insert(d.aptitude);
        comprehensions.insert(d.comprehension);
        EXPECT_GE(d.intelligence, 1);
        EXPECT_LE(d.intelligence, 200);
        EXPECT_GE(d.charm, 1);
        EXPECT_LE(d.charm, 200);
        EXPECT_GE(d.loyalty, 1);
        EXPECT_LE(d.loyalty, 100);
        EXPECT_GE(d.lifespan, 1);
        EXPECT_GE(d.hpVariance, -50);
        EXPECT_LE(d.hpVariance, 50);
        EXPECT_GE(d.speedVariance, -50);
        EXPECT_LE(d.speedVariance, 50);
    }
    // 数量全档可达（0-5）
    EXPECT_EQ(std::vector<int32_t>({0, 1, 2, 3, 4, 5}),
              std::vector<int32_t>(counts.begin(), counts.end()));
    // 资质永不等于哨兵 50
    EXPECT_EQ(0, aptitudes.count(50));
    // 五灵根悟性/资质 ∈ [1, 20]
    for (int32_t i = 0; i < 100; ++i) {
        auto rng = DeterministicRng::fromSeed(555 + i);
        DiscipleCreationSeed s = kSeed();
        s.id = "five-root-" + std::to_string(i);
        s.spiritRootType = "火,水,木,金,土";
        const auto d = createDisciple(s, rng);
        EXPECT_GE(d.comprehension, 1);
        EXPECT_LE(d.comprehension, 20);
        EXPECT_GE(d.aptitude, 1);
        EXPECT_LE(d.aptitude, 20);
        EXPECT_NE(d.aptitude, 50);
    }
    // 单灵根悟性 ∈ [80, 100]
    for (int32_t i = 0; i < 100; ++i) {
        auto rng = DeterministicRng::fromSeed(333 + i);
        DiscipleCreationSeed s = kSeed();
        s.id = "one-root-" + std::to_string(i);
        s.spiritRootType = "火";
        const auto d = createDisciple(s, rng);
        EXPECT_GE(d.comprehension, 80);
        EXPECT_LE(d.comprehension, 100);
    }
}

TEST(DiscipleFactory, RealmMaxAgeMattersForLifespan) {
    // 境界基准寿命参与计算：金丹（realm 7, maxAge 200）且无加成 ⇒ 200
    auto rng = DeterministicRng::fromSeed(20260901);
    DiscipleCreationSeed s = kSeed();
    s.id = "realm7";
    s.realm = 7;
    const auto d = createDisciple(s, rng);
    EXPECT_EQ(200, d.lifespan);
}

}  // namespace
